### DISCUSSION TITLE
Use nested foreach Loops in Twig Template

### DISCUSSION THREAD

**Discussion by ThePocky** [symfony/symfony#42886] [Q&A]
Hello,

in my Symfony project I use a Spotify API to display tracks and albums of a searched artist. In my SearchController I have now found a working solution. I solved this with nested foreach loops. But now I want to process this solution in my Twig template.

This is my function. It works so far and gives me all search results with the respective album and its title. I would like to use the search results like in this function in a Twig template. Unfortunately I'm currently a bit stuck, I can't find a solution to use this foreach loop in twig.

```
    public function index(SpotifyWebAPI $api): Response
    {

        $artists = json_decode(json_encode($api->search('ariana grande', 'artist')), true);

        foreach($artists['artists']['items'] as $key => $artist) {

            $alben = json_decode(json_encode($api->getArtistAlbums($artist['id'])), true);

            echo "<ul><li><b>".$artist['name']."</b><br>";

                    foreach($alben['items'] as $key => $album) {

                        $tracks = json_decode(json_encode($api->getAlbumTracks($album['id'])), true);

                        echo "<ul><li><i>".$album['name']."</i><br>";
        
                            foreach($tracks['items'] as $key => $track) {

                                echo "<ul><li>".$track['name']."<br></li></ul>";
        
                            }
        
                        echo "</li></ul>";
                    }

            echo "</li></ul>";

        }       

        return $this->render('search/index.html.twig');
    }
```

I hope someone can give me a tip on how to transfer these foreach loops into a twig template.

Basically I have tried a few solutions in the Twig template. The problem currently is that when I searched for the artist, I get the name and ID respectively. With the ID of the artist I can only search further and display the albums. This then continues, if I want to display the tracks, I need the ID of the respective album.

Thank You!

Greetings, Björn

**Comment by stof** [comment]
Well, Twig also allows nesting its `{% for %}` loops, so the way to do it is the same than in PHP: nest the loops.

**Comment by t0wqa** [comment]
Hi.

Try to separate logic of fetching data from logic of rendering view.

What, for example, can be done (pseudo-code):

first, we build array of all data, that can be passed to twig template afterwards:
```
$result = [];

$artists = json_decode(json_encode($api->search('ariana grande', 'artist')), `true);
foreach ($artists as $artist) {
    $artistId = $artist['id'];
    $albums =  json_decode(json_encode($api->getArtistAlbums($artistId)), true);

    $result[$artistId] = ['info' => $artist, 'albums' => $albums'];
    
    foreach ($albums as $album) {
        $albumId = $album['id'];
        $tracks = json_decode(json_encode($api->getAlbumTracks($albumId)), true);

        $result[$artistId]['albums'][$albumId] = ['info' => $album, 'tracks' => $tracks];
    }
}
```
Then, we pass this array to twig and easily render it using nested loops:
```
<ul>
{% for artist in result %}
    <li>
        <b>{{ artist['info']['title'] }}</b>
        <ul>
         {% for album in artist['albums'] %}
             <li>
                 <i>{{ album['info']['name'] }}</i>
                 <ul>
                  {% for track in album['tracks'] %}
                      <li>{{ track['name'] }}</li>
                  {% endfor %}
                 </ul>
             </li>
         {% endfor %}
        </ul>
    </li>
{% endfor %}
</ul>
```
I have not tested the code (as I mentioned, it is pseudo-code), but I hope you catch the general idea of what should be done

**Reply by ThePocky** [reply]
Hey,

Thank you very much, that helped me a lot. However, I am still stuck with only one problem. I have changed the code as follows, now the artists and the albums are displayed. However, the titles are not yet output, I always get the following error:

`Key "tracks" for array with keys "album_group, album_type, artists, available_markets, external_urls, href, id, images, name, release_date, release_date_precision, total_tracks, type, uri" does not exist.`

((Information: I have set the limit of the artist and the album result to 1.))

```
public function index(SpotifyWebAPI $api, $query, $type): Response
    {

        $result = [];

        $artists = json_decode(json_encode($api->search($query, $type, [
            'limit' => 1
        ])), true);
        foreach ($artists['artists']['items'] as $artist) {
            $artistId = $artist['id'];
            $albums =  json_decode(json_encode($api->getArtistAlbums($artistId, [
                'limit' => 1
            ])), true);
        
            $result[$artistId] = ['info' => $artist, 'albums' => $albums];
            
            foreach ($albums['items'] as $album) {
                $albumId = $album['id'];
                $tracks = json_decode(json_encode($api->getAlbumTracks($albumId)), true);
        
                $result[$artistId]['albums'][$albumId] = ['info' => $album, 'tracks' => $tracks];
            }
        }

        

        return $this->render('search/index.html.twig', [
            'dump' => print_r($result, true),
            'result' => $result
        ]);
    }
```

```
<ul>
{% for key,artist in result %}
    <li>
        <b>{{ artist['info']['name'] }}</b>
        <ul>
         {% for key,album in artist['albums']['items'] %}
             <li>
                 <i>{{ album['name'] }}</i>
                 <ul>
                  {% for key,track in album['tracks'] %}
                      <li>{{ track['name'] }}</li>
                  {% endfor %}
                 </ul>
             </li>
         {% endfor %}
        </ul>
    </li>
{% endfor %}
</ul>
```

Here I dumped the array to see which layer I need to work in. Unfortunately, I'm a bit stuck there. Maybe more than one eye is seeing the error. 

```Array
(
    [66CXWjxzNUsdJxJ2JdwvnR] => Array
        (
            [info] => Array
                (
                    [external_urls] => Array
                        (
                            [spotify] => https://open.spotify.com/artist/66CXWjxzNUsdJxJ2JdwvnR
                        )

                    [followers] => Array
                        (
                            [href] => 
                            [total] => 68332281
                        )

                    [genres] => Array
                        (
                            [0] => dance pop
                            [1] => pop
                            [2] => post-teen pop
                        )

                    [href] => https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR
                    [id] => 66CXWjxzNUsdJxJ2JdwvnR
                    [images] => Array
                        (
                            [0] => Array
                                (
                                    [height] => 640
                                    [url] => https://i.scdn.co/image/ab6761610000e5ebcdce7620dc940db079bf4952
                                    [width] => 640
                                )

                            [1] => Array
                                (
                                    [height] => 320
                                    [url] => https://i.scdn.co/image/ab67616100005174cdce7620dc940db079bf4952
                                    [width] => 320
                                )

                            [2] => Array
                                (
                                    [height] => 160
                                    [url] => https://i.scdn.co/image/ab6761610000f178cdce7620dc940db079bf4952
                                    [width] => 160
                                )

                        )

                    [name] => Ariana Grande
                    [popularity] => 94
                    [type] => artist
                    [uri] => spotify:artist:66CXWjxzNUsdJxJ2JdwvnR
                )

            [albums] => Array
                (
                    [href] => https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR/albums?offset=0&limit=1&include_groups=album,single,compilation,appears_on
                    [items] => Array
                        (
                            [0] => Array
                                (
                                    [album_group] => album
                                    [album_type] => album
                                    [artists] => Array
                                        (
                                            [0] => Array
                                                (
                                                    [external_urls] => Array
                                                        (
                                                            [spotify] => https://open.spotify.com/artist/66CXWjxzNUsdJxJ2JdwvnR
                                                        )

                                                    [href] => https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR
                                                    [id] => 66CXWjxzNUsdJxJ2JdwvnR
                                                    [name] => Ariana Grande
                                                    [type] => artist
                                                    [uri] => spotify:artist:66CXWjxzNUsdJxJ2JdwvnR
                                                )

                                        )

                                    [available_markets] => Array
                                        (
                                            [0] => AD
                                            [1] => AE
                                            [2] => AG
                                            [3] => AL
                                            [4] => AM
                                            [5] => AO
                                            [6] => AR
                                            [7] => AT
                                            [8] => AU
                                            [9] => AZ
                                            [10] => BA
                                            [11] => BB
                                            [12] => BD
                                            [13] => BE
                                            [14] => BF
                                            [15] => BG
                                            [16] => BH
                                            [17] => BI
                                            [18] => BJ
                                            [19] => BN
                                            [20] => BO
                                            [21] => BR
                                            [22] => BS
                                            [23] => BT
                                            [24] => BW
                                            [25] => BY
                                            [26] => BZ
                                            [27] => CA
                                            [28] => CH
                                            [29] => CI
                                            [30] => CL
                                            [31] => CM
                                            [32] => CO
                                            [33] => CR
                                            [34] => CV
                                            [35] => CW
                                            [36] => CY
                                            [37] => CZ
                                            [38] => DE
                                            [39] => DJ
                                            [40] => DK
                                            [41] => DM
                                            [42] => DO
                                            [43] => DZ
                                            [44] => EC
                                            [45] => EE
                                            [46] => EG
                                            [47] => ES
                                            [48] => FI
                                            [49] => FJ
                                            [50] => FM
                                            [51] => FR
                                            [52] => GA
                                            [53] => GB
                                            [54] => GD
                                            [55] => GE
                                            [56] => GH
                                            [57] => GM
                                            [58] => GN
                                            [59] => GQ
                                            [60] => GR
                                            [61] => GT
                                            [62] => GW
                                            [63] => GY
                                            [64] => HK
                                            [65] => HN
                                            [66] => HR
                                            [67] => HT
                                            [68] => HU
                                            [69] => ID
                                            [70] => IE
                                            [71] => IL
                                            [72] => IN
                                            [73] => IS
                                            [74] => IT
                                            [75] => JM
                                            [76] => JO
                                            [77] => JP
                                            [78] => KE
                                            [79] => KG
                                            [80] => KH
                                            [81] => KI
                                            [82] => KM
                                            [83] => KN
                                            [84] => KR
                                            [85] => KW
                                            [86] => KZ
                                            [87] => LA
                                            [88] => LB
                                            [89] => LC
                                            [90] => LI
                                            [91] => LK
                                            [92] => LR
                                            [93] => LS
                                            [94] => LT
                                            [95] => LU
                                            [96] => LV
                                            [97] => MA
                                            [98] => MC
                                            [99] => MD
                                            [100] => ME
                                            [101] => MG
                                            [102] => MH
                                            [103] => MK
                                            [104] => ML
                                            [105] => MN
                                            [106] => MO
                                            [107] => MR
                                            [108] => MT
                                            [109] => MU
                                            [110] => MV
                                            [111] => MW
                                            [112] => MX
                                            [113] => MY
                                            [114] => MZ
                                            [115] => NA
                                            [116] => NE
                                            [117] => NG
                                            [118] => NI
                                            [119] => NL
                                            [120] => NO
                                            [121] => NP
                                            [122] => NR
                                            [123] => NZ
                                            [124] => OM
                                            [125] => PA
                                            [126] => PE
                                            [127] => PG
                                            [128] => PH
                                            [129] => PK
                                            [130] => PL
                                            [131] => PS
                                            [132] => PT
                                            [133] => PW
                                            [134] => PY
                                            [135] => QA
                                            [136] => RO
                                            [137] => RS
                                            [138] => RU
                                            [139] => RW
                                            [140] => SA
                                            [141] => SB
                                            [142] => SC
                                            [143] => SE
                                            [144] => SG
                                            [145] => SI
                                            [146] => SK
                                            [147] => SL
                                            [148] => SM
                                            [149] => SN
                                            [150] => SR
                                            [151] => ST
                                            [152] => SV
                                            [153] => SZ
                                            [154] => TD
                                            [155] => TG
                                            [156] => TH
                                            [157] => TL
                                            [158] => TN
                                            [159] => TO
                                            [160] => TR
                                            [161] => TT
                                            [162] => TV
                                            [163] => TW
                                            [164] => TZ
                                            [165] => UA
                                            [166] => UG
                                            [167] => US
                                            [168] => UY
                                            [169] => UZ
                                            [170] => VC
                                            [171] => VN
                                            [172] => VU
                                            [173] => WS
                                            [174] => XK
                                            [175] => ZA
                                            [176] => ZM
                                            [177] => ZW
                                        )

                                    [external_urls] => Array
                                        (
                                            [spotify] => https://open.spotify.com/album/74vajFwEwXJ61OW1DKSPEa
                                        )

                                    [href] => https://api.spotify.com/v1/albums/74vajFwEwXJ61OW1DKSPEa
                                    [id] => 74vajFwEwXJ61OW1DKSPEa
                                    [images] => Array
                                        (
                                            [0] => Array
                                                (
                                                    [height] => 640
                                                    [url] => https://i.scdn.co/image/ab67616d0000b2739508fb7ca2eedc0d98b9139f
                                                    [width] => 640
                                                )

                                            [1] => Array
                                                (
                                                    [height] => 300
                                                    [url] => https://i.scdn.co/image/ab67616d00001e029508fb7ca2eedc0d98b9139f
                                                    [width] => 300
                                                )

                                            [2] => Array
                                                (
                                                    [height] => 64
                                                    [url] => https://i.scdn.co/image/ab67616d000048519508fb7ca2eedc0d98b9139f
                                                    [width] => 64
                                                )

                                        )

                                    [name] => Positions (Deluxe)
                                    [release_date] => 2021-02-19
                                    [release_date_precision] => day
                                    [total_tracks] => 19
                                    [type] => album
                                    [uri] => spotify:album:74vajFwEwXJ61OW1DKSPEa
                                )

                        )

                    [limit] => 1
                    [next] => https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR/albums?offset=1&limit=1&include_groups=album,single,compilation,appears_on
                    [offset] => 0
                    [previous] => 
                    [total] => 602
                    [74vajFwEwXJ61OW1DKSPEa] => Array
                        (
                            [info] => Array
                                (
                                    [album_group] => album
                                    [album_type] => album
                                    [artists] => Array
                                        (
                                            [0] => Array
                                                (
                                                    [external_urls] => Array
                                                        (
                                                            [spotify] => https://open.spotify.com/artist/66CXWjxzNUsdJxJ2JdwvnR
                                                        )

                                                    [href] => https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR
                                                    [id] => 66CXWjxzNUsdJxJ2JdwvnR
                                                    [name] => Ariana Grande
                                                    [type] => artist
                                                    [uri] => spotify:artist:66CXWjxzNUsdJxJ2JdwvnR
                                                )

                                        )

                                    [available_markets] => Array
                                        (
                                            [0] => AD
                                            [1] => AE
                                            [2] => AG
                                            [3] => AL
                                            [4] => AM
                                            [5] => AO
                                            [6] => AR
                                            [7] => AT
                                            [8] => AU
                                            [9] => AZ
                                            [10] => BA
                                            [11] => BB
                                            [12] => BD
                                            [13] => BE
                                            [14] => BF
                                            [15] => BG
                                            [16] => BH
                                            [17] => BI
                                            [18] => BJ
                                            [19] => BN
                                            [20] => BO
                                            [21] => BR
                                            [22] => BS
                                            [23] => BT
                                            [24] => BW
                                            [25] => BY
                                            [26] => BZ
                                            [27] => CA
                                            [28] => CH
                                            [29] => CI
                                            [30] => CL
                                            [31] => CM
                                            [32] => CO
                                            [33] => CR
                                            [34] => CV
                                            [35] => CW
                                            [36] => CY
                                            [37] => CZ
                                            [38] => DE
                                            [39] => DJ
                                            [40] => DK
                                            [41] => DM
                                            [42] => DO
                                            [43] => DZ
                                            [44] => EC
                                            [45] => EE
                                            [46] => EG
                                            [47] => ES
                                            [48] => FI
                                            [49] => FJ
                                            [50] => FM
                                            [51] => FR
                                            [52] => GA
                                            [53] => GB
                                            [54] => GD
                                            [55] => GE
                                            [56] => GH
                                            [57] => GM
                                            [58] => GN
                                            [59] => GQ
                                            [60] => GR
                                            [61] => GT
                                            [62] => GW
                                            [63] => GY
                                            [64] => HK
                                            [65] => HN
                                            [66] => HR
                                            [67] => HT
                                            [68] => HU
                                            [69] => ID
                                            [70] => IE
                                            [71] => IL
                                            [72] => IN
                                            [73] => IS
                                            [74] => IT
                                            [75] => JM
                                            [76] => JO
                                            [77] => JP
                                            [78] => KE
                                            [79] => KG
                                            [80] => KH
                                            [81] => KI
                                            [82] => KM
                                            [83] => KN
                                            [84] => KR
                                            [85] => KW
                                            [86] => KZ
                                            [87] => LA
                                            [88] => LB
                                            [89] => LC
                                            [90] => LI
                                            [91] => LK
                                            [92] => LR
                                            [93] => LS
                                            [94] => LT
                                            [95] => LU
                                            [96] => LV
                                            [97] => MA
                                            [98] => MC
                                            [99] => MD
                                            [100] => ME
                                            [101] => MG
                                            [102] => MH
                                            [103] => MK
                                            [104] => ML
                                            [105] => MN
                                            [106] => MO
                                            [107] => MR
                                            [108] => MT
                                            [109] => MU
                                            [110] => MV
                                            [111] => MW
                                            [112] => MX
                                            [113] => MY
                                            [114] => MZ
                                            [115] => NA
                                            [116] => NE
                                            [117] => NG
                                            [118] => NI
                                            [119] => NL
                                            [120] => NO
                                            [121] => NP
                                            [122] => NR
                                            [123] => NZ
                                            [124] => OM
                                            [125] => PA
                                            [126] => PE
                                            [127] => PG
                                            [128] => PH
                                            [129] => PK
                                            [130] => PL
                                            [131] => PS
                                            [132] => PT
                                            [133] => PW
                                            [134] => PY
                                            [135] => QA
                                            [136] => RO
                                            [137] => RS
                                            [138] => RU
                                            [139] => RW
                                            [140] => SA
                                            [141] => SB
                                            [142] => SC
                                            [143] => SE
                                            [144] => SG
                                            [145] => SI
                                            [146] => SK
                                            [147] => SL
                                            [148] => SM
                                            [149] => SN
                                            [150] => SR
                                            [151] => ST
                                            [152] => SV
                                            [153] => SZ
                                            [154] => TD
                                            [155] => TG
                                            [156] => TH
                                            [157] => TL
                                            [158] => TN
                                            [159] => TO
                                            [160] => TR
                                            [161] => TT
                                            [162] => TV
                                            [163] => TW
                                            [164] => TZ
                                            [165] => UA
                                            [166] => UG
                                            [167] => US
                                            [168] => UY
                                            [169] => UZ
                                            [170] => VC
                                            [171] => VN
                                            [172] => VU
                                            [173] => WS
                                            [174] => XK
                                            [175] => ZA
                                            [176] => ZM
                                            [177] => ZW
                                        )

                                    [external_urls] => Array
                                        (
                                            [spotify] => https://open.spotify.com/album/74vajFwEwXJ61OW1DKSPEa
                                        )

                                    [href] => https://api.spotify.com/v1/albums/74vajFwEwXJ61OW1DKSPEa
                                    [id] => 74vajFwEwXJ61OW1DKSPEa
                                    [images] => Array
                                        (
                                            [0] => Array
                                                (
                                                    [height] => 640
                                                    [url] => https://i.scdn.co/image/ab67616d0000b2739508fb7ca2eedc0d98b9139f
                                                    [width] => 640
                                                )

                                            [1] => Array
                                                (
                                                    [height] => 300
                                                    [url] => https://i.scdn.co/image/ab67616d00001e029508fb7ca2eedc0d98b9139f
                                                    [width] => 300
                                                )

                                            [2] => Array
                                                (
                                                    [height] => 64
                                                    [url] => https://i.scdn.co/image/ab67616d000048519508fb7ca2eedc0d98b9139f
                                                    [width] => 64
                                                )

                                        )

                                    [name] => Positions (Deluxe)
                                    [release_date] => 2021-02-19
                                    [release_date_precision] => day
                                    [total_tracks] => 19
                                    [type] => album
                                    [uri] => spotify:album:74vajFwEwXJ61OW1DKSPEa
                                )

                            [tracks] => Array
                                (
                                    [href] => https://api.spotify.com/v1/albums/74vajFwEwXJ61OW1DKSPEa/tracks?offset=0&limit=1
                                    [items] => Array
                                        (
                                            [0] => Array
                                                (
                                                    [artists] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [external_urls] => Array
                                                                        (
                                                                            [spotify] => https://open.spotify.com/artist/66CXWjxzNUsdJxJ2JdwvnR
                                                                        )

                                                                    [href] => https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR
                                                                    [id] => 66CXWjxzNUsdJxJ2JdwvnR
                                                                    [name] => Ariana Grande
                                                                    [type] => artist
                                                                    [uri] => spotify:artist:66CXWjxzNUsdJxJ2JdwvnR
                                                                )

                                                        )

                                                    [available_markets] => Array
                                                        (
                                                            [0] => AD
                                                            [1] => AE
                                                            [2] => AG
                                                            [3] => AL
                                                            [4] => AM
                                                            [5] => AO
                                                            [6] => AR
                                                            [7] => AT
                                                            [8] => AU
                                                            [9] => AZ
                                                            [10] => BA
                                                            [11] => BB
                                                            [12] => BD
                                                            [13] => BE
                                                            [14] => BF
                                                            [15] => BG
                                                            [16] => BH
                                                            [17] => BI
                                                            [18] => BJ
                                                            [19] => BN
                                                            [20] => BO
                                                            [21] => BR
                                                            [22] => BS
                                                            [23] => BT
                                                            [24] => BW
                                                            [25] => BY
                                                            [26] => BZ
                                                            [27] => CA
                                                            [28] => CH
                                                            [29] => CI
                                                            [30] => CL
                                                            [31] => CM
                                                            [32] => CO
                                                            [33] => CR
                                                            [34] => CV
                                                            [35] => CW
                                                            [36] => CY
                                                            [37] => CZ
                                                            [38] => DE
                                                            [39] => DJ
                                                            [40] => DK
                                                            [41] => DM
                                                            [42] => DO
                                                            [43] => DZ
                                                            [44] => EC
                                                            [45] => EE
                                                            [46] => EG
                                                            [47] => ES
                                                            [48] => FI
                                                            [49] => FJ
                                                            [50] => FM
                                                            [51] => FR
                                                            [52] => GA
                                                            [53] => GB
                                                            [54] => GD
                                                            [55] => GE
                                                            [56] => GH
                                                            [57] => GM
                                                            [58] => GN
                                                            [59] => GQ
                                                            [60] => GR
                                                            [61] => GT
                                                            [62] => GW
                                                            [63] => GY
                                                            [64] => HK
                                                            [65] => HN
                                                            [66] => HR
                                                            [67] => HT
                                                            [68] => HU
                                                            [69] => ID
                                                            [70] => IE
                                                            [71] => IL
                                                            [72] => IN
                                                            [73] => IS
                                                            [74] => IT
                                                            [75] => JM
                                                            [76] => JO
                                                            [77] => JP
                                                            [78] => KE
                                                            [79] => KG
                                                            [80] => KH
                                                            [81] => KI
                                                            [82] => KM
                                                            [83] => KN
                                                            [84] => KR
                                                            [85] => KW
                                                            [86] => KZ
                                                            [87] => LA
                                                            [88] => LB
                                                            [89] => LC
                                                            [90] => LI
                                                            [91] => LK
                                                            [92] => LR
                                                            [93] => LS
                                                            [94] => LT
                                                            [95] => LU
                                                            [96] => LV
                                                            [97] => MA
                                                            [98] => MC
                                                            [99] => MD
                                                            [100] => ME
                                                            [101] => MG
                                                            [102] => MH
                                                            [103] => MK
                                                            [104] => ML
                                                            [105] => MN
                                                            [106] => MO
                                                            [107] => MR
                                                            [108] => MT
                                                            [109] => MU
                                                            [110] => MV
                                                            [111] => MW
                                                            [112] => MX
                                                            [113] => MY
                                                            [114] => MZ
                                                            [115] => NA
                                                            [116] => NE
                                                            [117] => NG
                                                            [118] => NI
                                                            [119] => NL
                                                            [120] => NO
                                                            [121] => NP
                                                            [122] => NR
                                                            [123] => NZ
                                                            [124] => OM
                                                            [125] => PA
                                                            [126] => PE
                                                            [127] => PG
                                                            [128] => PH
                                                            [129] => PK
                                                            [130] => PL
                                                            [131] => PS
                                                            [132] => PT
                                                            [133] => PW
                                                            [134] => PY
                                                            [135] => QA
                                                            [136] => RO
                                                            [137] => RS
                                                            [138] => RU
                                                            [139] => RW
                                                            [140] => SA
                                                            [141] => SB
                                                            [142] => SC
                                                            [143] => SE
                                                            [144] => SG
                                                            [145] => SI
                                                            [146] => SK
                                                            [147] => SL
                                                            [148] => SM
                                                            [149] => SN
                                                            [150] => SR
                                                            [151] => ST
                                                            [152] => SV
                                                            [153] => SZ
                                                            [154] => TD
                                                            [155] => TG
                                                            [156] => TH
                                                            [157] => TL
                                                            [158] => TN
                                                            [159] => TO
                                                            [160] => TR
                                                            [161] => TT
                                                            [162] => TV
                                                            [163] => TW
                                                            [164] => TZ
                                                            [165] => UA
                                                            [166] => UG
                                                            [167] => US
                                                            [168] => UY
                                                            [169] => UZ
                                                            [170] => VC
                                                            [171] => VN
                                                            [172] => VU
                                                            [173] => WS
                                                            [174] => XK
                                                            [175] => ZA
                                                            [176] => ZM
                                                            [177] => ZW
                                                        )

                                                    [disc_number] => 1
                                                    [duration_ms] => 157710
                                                    [explicit] => 1
                                                    [external_urls] => Array
                                                        (
                                                            [spotify] => https://open.spotify.com/track/12sliObzVzbLbAlDb8eChf
                                                        )

                                                    [href] => https://api.spotify.com/v1/tracks/12sliObzVzbLbAlDb8eChf
                                                    [id] => 12sliObzVzbLbAlDb8eChf
                                                    [is_local] => 
                                                    [name] => shut up
                                                    [preview_url] => 
                                                    [track_number] => 1
                                                    [type] => track
                                                    [uri] => spotify:track:12sliObzVzbLbAlDb8eChf
                                                )

                                        )

                                    [limit] => 1
                                    [next] => https://api.spotify.com/v1/albums/74vajFwEwXJ61OW1DKSPEa/tracks?offset=1&limit=1
                                    [offset] => 0
                                    [previous] => 
                                    [total] => 19
                                )

                        )

                )

        )

)
```

Thank You.

Greetings, Björn

**Reply by t0wqa** [reply]
Hi!

I think we should simplify the way we build $result array. As we don't need all the data of API, but only names and titles, let's try the following:

In your controller (replace API calls with the correct ones from your code):

```
public function indexAction()
    {
        $result = [];

        $artists = json_decode(json_encode($this->getArtists()), true);
        foreach ($artists['items'] as $artist) {
            $artistId = $artist['id'];

            $artistResult = [
                'id' => $artistId,
                'name' => $artist['name'],
                'albums' => []
            ];

            $albums = json_decode(json_encode($this->getArtistAlbums($artistId)), true);
            foreach ($albums['items'] as $album) {
                $albumId = $album['id'];

                $albumResult = [
                    'id' => $albumId,
                    'name' => $album['name'],
                    'tracks' => []
                ];

                $tracks = json_decode(json_encode($this->getAlbumTracks($albumId)), true);
                foreach ($tracks['items'] as $track) {
                    $trackId = $track['id'];

                    $trackResult = [
                        'id' => $trackId,
                        'name' => $track['name']
                    ];

                    $albumResult['tracks'][] = $trackResult;
                }
                $artistResult['albums'][] = $albumResult;
            }
            $result[] = $artistResult;
        }

        return $this->render('Test/songs.html.twig', [
            'result' => $result
        ]);
    }
```

In twig:

```
<div>
    <ul>
        {% for artist in result %}
            <li>
                <b>{{ artist['name'] }}</b>
                <ul>
                    {% for album in artist['albums'] %}
                        <i>{{ album['name'] }}</i>
                        <ul>
                            {% for track in album['tracks'] %}
                                <li>{{ track['name'] }}</li>
                            {% endfor %}
                        </ul>
                    {% endfor %}
                </ul>
            </li>
        {% endfor %}
    </ul>
</div>
```

Also, it is a good practice to move the code which builds $result array to some service, so that controller method could use it instead of doing the operation itself. Also, to simplify debug in the future, it is best to use objects instead of array. But these are the next steps, for now let's make this at least work :)

**Reply by ThePocky** [reply]
It's incredible, it works! At the beginning I saw that the response of the SpotifyAPI are "stdobjects". Therefore I change at the fang with json_decode and json_encode a little back and forth to get an array. That was a little tinkering work until it worked partially. 

The basic function is established thanks to your help. Now I will experiment a little bit. 

I'm posting my entire controller here in case anyone comes across this post looking for a similar solution.

I used this Spotify Web API PHP: https://github.com/jwilsson/spotify-web-api-php

**_SearchController.php_**

```
<?php

namespace App\Controller;

use SpotifyWebAPI\SpotifyWebAPI;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;

class SearchController extends AbstractController
{
    /**
     * @Route("/search/{type}/{query}", name="search")
     */

    public function index(SpotifyWebAPI $api, $query, $type): Response
    {
        $result = [];

        $artists = json_decode(json_encode($api->search($query, $type)), true);
        foreach ($artists['artists']['items'] as $artist) {
            $artistId = $artist['id'];

            $artistResult = [
                'id' => $artistId,
                'name' => $artist['name'],
                'albums' => []
            ];

            $albums = json_decode(json_encode($api->getArtistAlbums($artistId)), true);
            foreach ($albums['items'] as $album) {
                $albumId = $album['id'];

                $albumResult = [
                    'id' => $albumId,
                    'name' => $album['name'],
                    'tracks' => []
                ];

                $tracks = json_decode(json_encode($api->getAlbumTracks($albumId)), true);
                foreach ($tracks['items'] as $track) {
                    $trackId = $track['id'];

                    $trackResult = [
                        'id' => $trackId,
                        'name' => $track['name']
                    ];

                    $albumResult['tracks'][] = $trackResult;
                }
                $artistResult['albums'][] = $albumResult;
            }
            $result[] = $artistResult;
        }

        return $this->render('search/index.html.twig', [
            'result' => $result
        ]);
    }
}
```

**_search/index.html.twig_**

```
{% extends 'base.html.twig' %}

{% block title %}SpotifySearch{% endblock %}

{% block body %}

<ul>
        {% for artist in result %}
            <li>
                <b>{{ artist['name'] }}</b>
                <ul>
                    {% for album in artist['albums'] %}
                        <i>{{ album['name'] }}</i>
                        <ul>
                            {% for track in album['tracks'] %}
                                <li>{{ track['name'] }}</li>
                            {% endfor %}
                        </ul>
                    {% endfor %}
                </ul>
            </li>
        {% endfor %}
    </ul>

{% endblock %}

```

Thank You so much for your help!

Greetings, Björn
